src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
-
-        private volatile RxRingBuffer scalarValueQueue = null;
-
-        /* protected by lock on MergeSubscriber instance */
-        private int missedEmitting = 0;
-        private boolean emitLock = false;
+    static final class MergeProducer<T> extends AtomicLong implements Producer {
+        /** */
+        private static final long serialVersionUID = -1214379189873595503L;
 
-        /**
-         * Using synchronized(this) for `emitLock` instead of ReentrantLock or AtomicInteger is faster when there is no contention.
-         * 
-         * <pre> {@code
-         * Using ReentrantLock:
-         * r.o.OperatorMergePerf.merge1SyncStreamOfN           1000  thrpt         5    44185.294     1295.565    ops/s
-         * 
-         * Using synchronized(this):
-         * r.o.OperatorMergePerf.merge1SyncStreamOfN           1000  thrpt         5    79715.981     3704.486    ops/s
-         * 
-         * Still slower though than allowing concurrency:
-         * r.o.OperatorMergePerf.merge1SyncStreamOfN           1000  thrpt         5   149331.046     4851.290    ops/s
-         * } </pre>
-         */
-
-        public MergeSubscriber(Subscriber<? super T> actual, boolean delayErrors) {
-            super(actual);
-            this.actual = actual;
-            this.mergeProducer = new MergeProducer<T>(this);
-            this.delayErrors = delayErrors;
-            // decoupled the subscription chain because we need to decouple and control backpressure
-            actual.add(this);
-            actual.setProducer(mergeProducer);
-        }
-
-        @Override
-        public void onStart() {
-            // we request backpressure so we can handle long-running Observables that are enqueueing, such as flatMap use cases
-            // we decouple the Producer chain while keeping the Subscription chain together (perf benefit) via super(actual)
-            request(RxRingBuffer.SIZE);
+        final MergeSubscriber<T> subscriber;
+        /** Indicates the producer mode: -1 no request, 0 unbounded fast, 1 bounded. */
+        volatile int mode;
+        @SuppressWarnings("rawtypes")
+        static final AtomicIntegerFieldUpdater<MergeProducer> MODE
+        = AtomicIntegerFieldUpdater.newUpdater(MergeProducer.class, "mode");
+        
+        public MergeProducer(MergeSubscriber<T> subscriber) {
+            this.subscriber = subscriber;
+            this.mode = -1;
         }
-
-        /*
-         * This is expected to be executed sequentially as per the Rx contract or it will not work.
-         */
+        
         @Override
-        public void onNext(Observable<? extends T> t) {
-            if (t instanceof ScalarSynchronousObservable) {
-                ScalarSynchronousObservable<? extends T> t2 = (ScalarSynchronousObservable<? extends T>)t;
-                handleScalarSynchronousObservable(t2);
-            } else {
-                if (t == null || isUnsubscribed()) {
+        public void request(long n) {
```

Comment:
This `request` method, and perhaps this entire `Producer` implementation seems to make sense as part of BackpressureUtils as it is mostly generic. The `subscriber.*` methods that are invoked could be a `BackpressureSubscriber` interface perhaps?

This is code that seems reusable for all of our backpressure aware operators.